x-pack/plugins/security_solution/server/endpoint/services/artifacts/manifest_manager/manifest_manager.ts

```diff
@@ -129,45 +138,44 @@ export class ManifestManager {
   }
 
   /**
-   * Builds an artifact (one per supported OS) based on the current
-   * state of exception-list-agnostic SOs.
+   * Search or get exceptions from the cached map by listId and OS and filter those by policyId/global
    */
-  protected async buildExceptionListArtifact(os: string): Promise<InternalArtifactCompleteSchema> {
-    return buildArtifact(
-      await getEndpointExceptionList({
-        elClient: this.exceptionListClient,
-        schemaVersion: this.schemaVersion,
+  protected async getCachedExceptions({
+    elClient,
+    listId,
+    os,
+    policyId,
+    schemaVersion,
+  }: {
+    elClient: ExceptionListClient;
+    listId: ArtifactListId;
+    os: string;
+    policyId?: string;
+    schemaVersion: string;
+  }) {
+    if (!this.cachedExceptionsListsByOs.has(`${listId}-${os}`)) {
+      const itemsByListId = await getAllItemsFromEndpointExceptionList({
+        elClient,
         os,
-      }),
-      this.schemaVersion,
-      os,
-      ArtifactConstants.GLOBAL_ALLOWLIST_NAME
-    );
-  }
-
-  /**
-   * Builds an array of artifacts (one per supported OS) based on the current
-   * state of exception-list-agnostic SOs.
-   *
-   * @returns {Promise<InternalArtifactCompleteSchema[]>} An array of uncompressed artifacts built from exception-list-agnostic SOs.
-   * @throws Throws/rejects if there are errors building the list.
-   */
-  protected async buildExceptionListArtifacts(): Promise<ArtifactsBuildResult> {
-    const defaultArtifacts: InternalArtifactCompleteSchema[] = [];
-    const policySpecificArtifacts: Record<string, InternalArtifactCompleteSchema[]> = {};
+        listId,
+      });
+      this.cachedExceptionsListsByOs.set(`${listId}-${os}`, itemsByListId);
+    }
 
-    for (const os of ArtifactConstants.SUPPORTED_OPERATING_SYSTEMS) {
-      defaultArtifacts.push(await this.buildExceptionListArtifact(os));
+    const allExceptionsByListId = this.cachedExceptionsListsByOs.get(`${listId}-${os}`);
+    if (!allExceptionsByListId) {
+      throw new InvalidInternalManifestError(`Error getting exceptions for ${listId}-${os}`);
     }
 
-    await iterateAllListItems(
-      (page) => this.listEndpointPolicyIds(page),
-      async (policyId) => {
-        policySpecificArtifacts[policyId] = defaultArtifacts;
-      }
-    );
+    const filter = (exception: ExceptionListItemSchema) =>
+      policyId
+        ? exception.tags.includes('policy:all') || exception.tags.includes(`policy:${policyId}`)
```

Comment:
So if there is a `policyId`, then we want to filter global and by policy exceptions, otherwise, we only want to filter the global ones. Am I missing something?

Reply:
ohhh. Ok. Sorry. I think I got it now.